### PR TITLE
rbd: add data pool support to trash purge

### PR DESCRIPTION
**Pull request by MahatiC**
Checks implicitly for datapool utilization of all images and removes
respective pool's trashed images until threshold value is hit.

Fixes: http://tracker.ceph.com/issues/22872
Signed-off-by: Mahati Chamarthy <mahati.chamarthy@intel.com>

### CODE REVIEW

**Comment by batrick** [comment]
retest this please

**Comment by MahatiC** [comment]
retest this please

**Comment by MahatiC** [comment]
retest this please

**Comment by tchaikov** [comment]
retest this please

File: src/test/cli/rbd/help.t [hunk 1]
```diff
@@ -1624,6 +1624,7 @@
   usage: rbd trash purge [--pool <pool>] [--no-progress] 
                          [--expired-before <expired-before>] 
                          [--threshold <threshold>] 
+                         [--data-pool <data-pool>]
```
Comment (tchaikov) [inline src/test/cli/rbd/help.t]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/rbd/help.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/rbd/help.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/rbd/help.t.err
@@ -1626,8 +1626,7 @@
   rbd help trash purge
   usage: rbd trash purge [--pool <pool>] [--no-progress] 
                          [--expired-before <expired-before>] 
-                         [--threshold <threshold>] 
-                         [--data-pool <data-pool>]
+                         [--threshold <threshold>] [--data-pool <data-pool>] 
                          <pool-name> 
   
   Remove all expired images from trash.
```

you might need to move "`[--data-pool <data-pool>`" up to the previous line.

**Comment by MahatiC** [comment]
@dillaman rbd trash purge seems to accept only one pool. And will only have access to trashed images (whose metadata is stored) in that particular pool. For example:

rbd ls rbdpool    
image1, image2, image3

datapool1 has {image1, otherimage1}
datapool2 has {image2, otherimage2, otherimage3}

rbd trash mv rbdpool/image1
rbd trash mv rbdpool/image2

On excuting: rbd trash purge rbdpool --threshold '0.02'

do you mean remove images in rbdpool until datapool1 & datapool2 both have used % of 2? In the above example datapool2 has images from a pool other than rbdpool. Not sure I understand when you say remove images from data pool.

**Comment by MahatiC** [comment]
@dillaman review please

File: src/tools/rbd/action/Trash.cc [hunk 2]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+        if (r < 0) {
+          data_pool = "<missing data pool " + stringify(data_pool_id) + ">";
+        } else {
+          data_pool = data_io_ctx.get_pool_name();
+          datapools[data_pool];
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: unnecessary

File: src/tools/rbd/action/Trash.cc [hunk 3]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
         return a.deferment_end_time < b.deferment_end_time;
       }
     );
 
-    uint64_t bytes_to_free = 0;
-    auto bytes_threshold = (uint64_t)(pool_total_bytes * 
-                           (pool_percent_used - threshold));
-    
-    librbd::Image curr_img;
+    std::string data_pool;
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: move to tightest scope

File: src/tools/rbd/action/Trash.cc [hunk 4]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+      r = utils::open_image_by_id(io_ctx, entry.id, true, &image);
       if(r < 0) continue;
+
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+        if (r < 0) {
+          data_pool = "<missing data pool " + stringify(data_pool_id) + ">";
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: unnecessary

File: src/tools/rbd/action/Trash.cc [hunk 5]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+        } else {
+          data_pool = data_io_ctx.get_pool_name();
+          datapools[data_pool];
+          datapools[data_pool].push_back(entry.id);
+        }
+      }
+    }
+    datapools[pool_name];
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: awkward -- might be clearer if you use ```datapools[pool_name] = {}``` or ```datapools.insert({pool_name}, {}})```

File: src/tools/rbd/action/Trash.cc [hunk 6]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+          data_pool = "<missing data pool " + stringify(data_pool_id) + ">";
+        } else {
+          data_pool = data_io_ctx.get_pool_name();
+          datapools[data_pool];
+          datapools[data_pool].push_back(entry.id);
+        }
+      }
+    }
+    datapools[pool_name]; 
+    uint64_t bytes_to_free;
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: initialize your variable

File: src/tools/rbd/action/Trash.cc [hunk 7]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+
+    for(uint8_t i = 0; i < arr.size(); ++i) {
+      json_spirit::mObject obj = arr[i].get_obj();
+      std::string name = obj.find("name")->second.get_str();
+      std::map<std::string, std::vector<std::string>>::iterator img = datapools.find(name);
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: ```auto img = ...```

File: src/tools/rbd/action/Trash.cc [hunk 8]
```diff
@@ -417,53 +417,103 @@ int execute_purge (const po::variables_map &vm,
...
+            
+        pool_total_bytes = stats["max_avail"].get_uint64() +
+                           stats["bytes_used"].get_uint64();
+          
+        auto bytes_threshold = (uint64_t)(pool_total_bytes * 
+                               (pool_percent_used - threshold));
+    
+        if(arr[i].get_obj()["name"] == pool_name){
+          librbd::Image curr_img;
+          for (const auto& entry : trash_entries) {
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Seems to be duplicating the code below. If you added the trash image ids for the base pool to ```datapools[pool_name]``` above, you could then just iterate through the image ids in the ```datapools[arr[i].get_obj()["name"]]``` to compute which images to remove.

**Comment by MahatiC** [comment]
@dillaman Thanks, updated.

File: src/tools/rbd/action/Trash.cc [hunk 9]
```diff
@@ -423,53 +423,79 @@ int execute_purge (const po::variables_map &vm,
...
+
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+        data_pool = data_io_ctx.get_pool_name();
+        datapools[data_pool].push_back(entry.id.c_str());
+      }
+      datapools[pool_name].push_back(entry.id.c_str());
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: On second thought -- I guess my suggestion would double-count an image's space in both its data pool and base pool. Therefore, the disk usage check for determining which images to delete should only be made in the data pool. Therefore, perhaps this should just be the ```else``` branch.

File: src/tools/rbd/action/Trash.cc [hunk 10]
```diff
@@ -423,53 +423,79 @@ int execute_purge (const po::variables_map &vm,
...
+      librbd::Image image;
+      std::string data_pool;
+      r = utils::open_image_by_id(io_ctx, entry.id, true, &image);
       if(r < 0) continue;
+
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
```
Comment (dillaman) [inline src/tools/rbd/action/Trash.cc]: Nit: log error message to ```std::cerr``` and skip image if ```r < 0````

File: src/tools/rbd/action/Trash.cc (hunk 9 above)
Comment (MahatiC) [reply src/tools/rbd/action/Trash.cc]: Okay. With your prev suggestion I reasoned that technically all images belong to the base pool anyway and should be fine if we delete (datapool) images from basepool (where datapool's threshold is what is required and basepool is higher). But not double-counting makes it easier to differentiate and the (trash purge) behavior more predictable. Will modify.

File: src/tools/rbd/action/Trash.cc (hunk 10 above)
Comment (MahatiC) [reply src/tools/rbd/action/Trash.cc]: Done.

File: src/tools/rbd/action/Trash.cc (hunk 9 above)
Comment (MahatiC) [reply src/tools/rbd/action/Trash.cc]: Done.

File: src/tools/rbd/action/Trash.cc [hunk 11]
```diff
@@ -423,53 +423,85 @@ int execute_purge (const po::variables_map &vm,
...
+        if (r < 0) {
+          std::cerr << "rbd: error accessing data pool" << std::endl;
+          continue;
+        }
+        data_pool = data_io_ctx.get_pool_name();
+        datapools[data_pool].push_back(entry.id.c_str());
+      }
+      else {
```
[the remark this answers is not in the document]
Comment (MahatiC) [reply src/tools/rbd/action/Trash.cc]: Done.

File: src/tools/rbd/action/Trash.cc [hunk 12]
```diff
@@ -423,53 +423,85 @@ int execute_purge (const po::variables_map &vm,
...
+      datapools[pool_name].push_back(entry.id.c_str());
```
[the remark this answers is not in the document]
Comment (MahatiC) [reply src/tools/rbd/action/Trash.cc]: Done.

**Pull request merged**